### PR TITLE
matrix-synapse: 0.22.1 -> 0.24.1

### PR DESCRIPTION
**Pull request by Ralith**
###### Motivation for this change
Upstream optimizations and fixes

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Ralith** [comment]
I don't know the first thing about python packaging; maybe there's a tidier way to handle affinity.

**Comment by eqyiel** [comment]
LGTM, but I think there's two unconventional things about the affinity expression: 
- use of `buildPythonPackage` instead of `buildPythonApplication` for a library (or is affinity meant to be called interactively?
- `fetchPypi` with `pname`, `version` and `sha256` attributes is preferred over `fetchurl`

**Comment by Ralith** [comment]
Thanks! Fixed.

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -105,6 +105,16 @@ in {
 
   aenum = callPackage ../development/python-modules/aenum { };
 
+  affinity = buildPythonPackage rec {
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: See the header of this file and the Nixpkgs manual for contributing Python packages.

**Review by FRidh — CHANGES_REQUESTED** [review]

**Comment by LnL7** [comment]
@grahamcofborg matrix-synapse

File: pkgs/development/python-modules/affinity/default.nix [hunk 2]
```diff
@@ -0,0 +1,11 @@
+{ buildPythonPackage, fetchPypi }:
+
+buildPythonPackage rec {
+  name = "${pname}-${version}";
+  pname = "affinity";
+  version = "0.1.0";
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "1i6j7kszvnzh5vh9k48cqwx2kzf73a6abgv9s6bf0j2zmfjl2wb6";
+  };
```
Comment (FRidh) [inline pkgs/development/python-modules/affinity/default.nix]: meta is missing

**Review by FRidh — CHANGES_REQUESTED** [review]

**Comment by Ralith** [comment]
Bumped version again.

**Comment by rycee** [comment]
Builds OK for me. Haven't tried running it but would very much like to see this in master.

@Ralith Any remaining issues that you feel blocks this?

**Review by fpletz — APPROVED** [review]

**Comment by Ralith** [comment]
AFAIK this is ready; I've been using it on a live system for a while now.

File: pkgs/development/python-modules/affinity/default.nix (hunk 2 above)
Comment (corngood) [reply pkgs/development/python-modules/affinity/default.nix]: @FRidh, this appears to be fixed.

**Pull request merged**